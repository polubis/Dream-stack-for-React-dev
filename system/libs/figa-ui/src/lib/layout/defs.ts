import type { ReactNode } from 'react';

interface LayoutProps {
  className?: string;
  children: ReactNode;
  header: ReactNode;
  full?: boolean;
  footer?: ReactNode;
}

export type { LayoutProps };
