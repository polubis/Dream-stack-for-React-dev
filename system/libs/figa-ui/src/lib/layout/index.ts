export * from './defs';
export * from './layout';
