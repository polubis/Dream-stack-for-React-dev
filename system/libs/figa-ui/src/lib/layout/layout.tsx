import { useScroll, useToggle } from '@system/figa-hooks';
import type { LayoutProps } from './defs';

import c from 'classnames';
import styled from 'styled-components';
import { column, row, streched } from '../shared';
import { tokens } from '../theme-provider';

const Container = styled.div`
  ${column()}
  padding-top: ${tokens.spacing[1000]};

  &.asided {
    .layout-content {
      padding: 0;
      display: flex;
      min-height: calc(100vh - ${tokens.spacing[1000]});
      max-width: 100vw;

      & > * {
        width: 100%;
      }

      & > *:first-child {
        border-right: ${tokens.spacing[25]} solid
          ${(props) => props.theme.nav.borderColor};
        background: ${(props) => props.theme.nav.bg};
        width: ${tokens.spacing[1000]};
        padding: ${tokens.spacing[350]} 0;
      }

      & > *:last-child {
        padding: ${tokens.spacing[350]} ${tokens.spacing[250]};
      }
    }

    &.opened {
      .layout-content {
        & > *:first-child {
          width: 300px;
          padding: ${tokens.spacing[350]} ${tokens.spacing[250]};
        }

        & > *:last-child {
          padding: ${tokens.spacing[350]} ${tokens.spacing[250]};
        }
      }
    }
  }

  .layout-content {
    padding: ${tokens.spacing[350]} ${tokens.spacing[250]};
    display: flex;
    min-height: calc(100vh - 100px);
    max-width: 100vw;

    & > * {
      width: 100%;
    }
  }

  &.off-padding {
    .layout-content {
      padding: 0;
    }

    &.asided {
      .layout-content {
        & > *:last-child {
          padding: 0;
        }
      }
    }
  }

  .layout-top-bar {
    ${row()}
    ${streched('fixed')}
        bottom: unset;
    border-bottom: ${tokens.spacing[25]} solid
      ${(props) => props.theme.nav.borderColor};
    background: ${(props) => props.theme.nav.bgWithOpacity};
    padding: 0 ${tokens.spacing[250]};
    z-index: ${tokens.z[200]};
    height: ${tokens.spacing[1000]};
    transform: translateY(0px);
    transition: 0.93s transform cubic-bezier(0.19, 1, 0.22, 1);

    &.out {
      transform: translateY(-${tokens.spacing[1000]});
    }
  }
`;

const Layout = ({
  className,
  children,
  topNav,
  footer,
  offPadding,
  sidebar,
}: LayoutProps) => {
  const toggler = useToggle({ opened: !!sidebar });
  const [state] = useScroll({ delay: 50 });

  return (
    <Container
      className={c(
        'layout',
        className,
        { asided: !!sidebar },
        { opened: toggler.opened },
        { 'off-padding': offPadding }
      )}
    >
      <header
        className={c('layout-top-bar', className, {
          out: state.is === 'progress' && state.value > 50,
        })}
      >
        {topNav}
      </header>
      <main className="layout-content">
        {sidebar && (
          <aside className="layout-content-sidebar">{sidebar(toggler)}</aside>
        )}
        {children}
      </main>
      <footer className="layout-footer">{footer}</footer>
    </Container>
  );
};

export { Layout };
