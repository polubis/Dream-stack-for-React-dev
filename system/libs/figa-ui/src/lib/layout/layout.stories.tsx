import type { Story, Meta } from '@storybook/react';
import type { LayoutProps } from './defs';

import { Layout } from './layout';
import { Logo } from '../logo';
import { Font } from '../font';
import { Nav } from '../nav';
import { TopNavItem } from '../top-nav-item';
import { PlusCircleIcon } from '../icon';

export default {
  component: Layout,
  title: 'Layout',
} as Meta;

const Header = () => (
  <Nav
    logo={<Logo />}
    actions={
      <>
        <TopNavItem active>
          Create <PlusCircleIcon />
        </TopNavItem>
        <TopNavItem active>
          Create <PlusCircleIcon />
        </TopNavItem>
      </>
    }
  >
    <a href="/">
      <TopNavItem active>
        Create <PlusCircleIcon />
      </TopNavItem>
    </a>
    <a href="/">
      <TopNavItem active>
        Create <PlusCircleIcon />
      </TopNavItem>
    </a>
    <a href="/">
      <TopNavItem active>
        Create <PlusCircleIcon />
      </TopNavItem>
    </a>
    <a href="/">
      <TopNavItem active>
        Create <PlusCircleIcon />
      </TopNavItem>
    </a>
  </Nav>
);

const BigText = (
  <Font variant="h6">
    . Why do we use it? It is a long established fact that a reader will be
    distracted by the readable content of a page when looking at its layout. The
    point of using Lorem Ipsum is that it has a more-or-less normal distribution
    of letters, as opposed to using 'Content here, content here', making it look
    like readable English. Many desktop publishing packages and web page editors
    now use Lorem Ipsum as their default model text, and a search for 'lorem
    ipsum' will uncover many web sites still in their infancy. Various versions
    have evolved over the years, sometimes by accident, sometimes on purpose
    (injected humour and the like). Where does it come from? Contrary to popular
    belief, Lorem Ipsum is not simply random text. It has roots in a piece of
    classical Latin literature from 45 BC, making it over 2000 years old.
    Richard McClintock, a Latin professor at Hampden-Sydney College in Virginia,
    looked up one of the more obscure Latin words, consectetur, from a Lorem
    Ipsum passage, and going through the cites of the word in classical
    literature, discovered the undoubtable source. Lorem Ipsum comes from
    sections 1.10.32 and 1.10.33 of "de Finibus Bonorum et Malorum" (The
    Extremes of Good and Evil) by Cicero, written in 45 BC. This book is a
    treatise on the theory of ethics, very popular during the Renaissance. The
    first line of Lorem Ipsum, "Lorem ipsum dolor sit amet..", comes from a line
    in section 1.10.32. The standard chunk of Lorem Ipsum used since the 1500s
    is reproduced below for those interested. Sections 1.10.32 and 1.10.33 from
    "de Finibus Bonorum et Malorum" by Cicero are also reproduced in their exact
    original form, accompanied by English versions from the 1914 translation by
    H. Rackham. Where can I get some? There are many variations of passages of
    Lorem Ipsum available, but the majority have suffered alteration in some
    form, by injected humour, or randomised words which don't look even slightly
    believable. If you are going to use a passage of Lorem Ipsum, you need to be
    sure there isn't anything embarrassing hidden in the middle of text. All the
    Lorem Ipsum generators on the Internet tend to repeat predefined chunks as
    necessary, making this the first true generator on the Internet. It uses a
    dictionary of over 200 Latin words, combined with a handful of model
    sentence structures, to generate Lorem Ipsum which looks reasonable. The
    generated Lorem Ipsum is therefore always free from repetition, injected
    humour, or non-characteristic words etc. 5 paragraphs words bytes lists
    Start with 'Lorem ipsum dolor sit amet...' . Why do we use it? It is a long
    established fact that a reader will be distracted by the readable content of
    a page when looking at its layout. The point of using Lorem Ipsum is that it
    has a more-or-less normal distribution of letters, as opposed to using
    'Content here, content here', making it look like readable English. Many
    desktop publishing packages and web page editors now use Lorem Ipsum as
    their default model text, and a search for 'lorem ipsum' will uncover many
    web sites still in their infancy. Various versions have evolved over the
    years, sometimes by accident, sometimes on purpose (injected humour and the
    like). Where does it come from? Contrary to popular belief, Lorem Ipsum is
    not simply random text. It has roots in a piece of classical Latin
    literature from 45 BC, making it over 2000 years old. Richard McClintock, a
    Latin professor at Hampden-Sydney College in Virginia, looked up one of the
    more obscure Latin words, consectetur, from a Lorem Ipsum passage, and going
    through the cites of the word in classical literature, discovered the
    undoubtable source. Lorem Ipsum comes from sections 1.10.32 and 1.10.33 of
    "de Finibus Bonorum et Malorum" (The Extremes of Good and Evil) by Cicero,
    written in 45 BC. This book is a treatise on the theory of ethics, very
    popular during the Renaissance. The first line of Lorem Ipsum, "Lorem ipsum
    dolor sit amet..", comes from a line in section 1.10.32. The standard chunk
    of Lorem Ipsum used since the 1500s is reproduced below for those
    interested. Sections 1.10.32 and 1.10.33 from "de Finibus Bonorum et
    Malorum" by Cicero are also reproduced in their exact original form,
    accompanied by English versions from the 1914 translation by H. Rackham.
    Where can I get some? There are many variations of passages of Lorem Ipsum
    available, but the majority have suffered alteration in some form, by
    injected humour, or randomised words which don't look even slightly
    believable. If you are going to use a passage of Lorem Ipsum, you need to be
    sure there isn't anything embarrassing hidden in the middle of text. All the
    Lorem Ipsum generators on the Internet tend to repeat predefined chunks as
    necessary, making this the first true generator on the Internet. It uses a
    dictionary of over 200 Latin words, combined with a handful of model
    sentence structures, to generate Lorem Ipsum which looks reasonable. The
    generated Lorem Ipsum is therefore always free from repetition, injected
    humour, or non-characteristic words etc. 5 paragraphs words bytes lists
    Start with 'Lorem ipsum dolor sit amet...' . Why do we use it? It is a long
    established fact that a reader will be distracted by the readable content of
    a page when looking at its layout. The point of using Lorem Ipsum is that it
    has a more-or-less normal distribution of letters, as opposed to using
    'Content here, content here', making it look like readable English. Many
    desktop publishing packages and web page editors now use Lorem Ipsum as
    their default model text, and a search for 'lorem ipsum' will uncover many
    web sites still in their infancy. Various versions have evolved over the
    years, sometimes by accident, sometimes on purpose (injected humour and the
    like). Where does it come from? Contrary to popular belief, Lorem Ipsum is
    not simply random text. It has roots in a piece of classical Latin
    literature from 45 BC, making it over 2000 years old. Richard McClintock, a
    Latin professor at Hampden-Sydney College in Virginia, looked up one of the
    more obscure Latin words, consectetur, from a Lorem Ipsum passage, and going
    through the cites of the word in classical literature, discovered the
    undoubtable source. Lorem Ipsum comes from sections 1.10.32 and 1.10.33 of
    "de Finibus Bonorum et Malorum" (The Extremes of Good and Evil) by Cicero,
    written in 45 BC. This book is a treatise on the theory of ethics, very
    popular during the Renaissance. The first line of Lorem Ipsum, "Lorem ipsum
    dolor sit amet..", comes from a line in section 1.10.32. The standard chunk
    of Lorem Ipsum used since the 1500s is reproduced below for those
    interested. Sections 1.10.32 and 1.10.33 from "de Finibus Bonorum et
    Malorum" by Cicero are also reproduced in their exact original form,
    accompanied by English versions from the 1914 translation by H. Rackham.
    Where can I get some? There are many variations of passages of Lorem Ipsum
    available, but the majority have suffered alteration in some form, by
    injected humour, or randomised words which don't look even slightly
    believable. If you are going to use a passage of Lorem Ipsum, you need to be
    sure there isn't anything embarrassing hidden in the middle of text. All the
    Lorem Ipsum generators on the Internet tend to repeat predefined chunks as
    necessary, making this the first true generator on the Internet. It uses a
    dictionary of over 200 Latin words, combined with a handful of model
    sentence structures, to generate Lorem Ipsum which looks reasonable. The
    generated Lorem Ipsum is therefore always free from repetition, injected
    humour, or non-characteristic words etc. 5 paragraphs words bytes lists
    Start with 'Lorem ipsum dolor sit amet...'
  </Font>
);

const Template: Story<LayoutProps> = (props) => <Layout {...props} />;

export const Default = Template.bind({});
Default.args = {
  topNav: <Header />,
  children: (
    <div>
      <Font variant="h6">Custom header</Font>
    </div>
  ),
  footer: <div>Footer</div>,
};

export const ALotOfContent = Template.bind({});
ALotOfContent.args = {
  topNav: <Header />,
  children: <div>{BigText}</div>,
  footer: <div>Footer</div>,
};

export const WithoutFooter = Template.bind({});
WithoutFooter.args = {
  topNav: <Header />,
  children: <div>{BigText}</div>,
};

export const WhenPageIsFull = Template.bind({});
WhenPageIsFull.args = {
  topNav: <Header />,
  children: <div style={{ background: 'red' }}>Text</div>,
};

export const WhenPageIsFullWithFooter = Template.bind({});
WhenPageIsFullWithFooter.args = {
  topNav: <Header />,
  children: <div style={{ background: 'red' }}>Text</div>,
  footer: <div>Footer</div>,
};

export const WithoutPadding = Template.bind({});
WithoutPadding.args = {
  topNav: <Header />,
  children: <div style={{ background: 'red' }}>Text</div>,
  footer: <div>Footer</div>,
  offPadding: true,
};

export const AsidedWithoutPadding = Template.bind({});
AsidedWithoutPadding.args = {
  topNav: <Header />,
  children: (
    <div>
      <Font variant="h2">The content</Font>
    </div>
  ),
  footer: <div>Footer</div>,
  sidebar: (toggler) => (
    <aside onClick={toggler.toggle}>
      <Font variant="h5">The sidebar</Font>
    </aside>
  ),
  offPadding: true,
};

export const Asided = Template.bind({});
Asided.args = {
  topNav: <Header />,
  children: (
    <div>
      <Font variant="h2">The content</Font>
    </div>
  ),
  footer: <div>Footer</div>,
  sidebar: (toggler) => (
    <aside onClick={toggler.toggle}>
      <Font variant="h5">The sidebar</Font>
    </aside>
  ),
};
