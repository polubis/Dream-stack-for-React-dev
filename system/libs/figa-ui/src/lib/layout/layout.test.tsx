import { fireEvent, render, screen } from '@testing-library/react';

import { Layout } from './layout';

describe('User is able to use layout when', () => {
  it('[FRAGILE] renders with default setup', () => {
    const { asFragment } = render(
      <Layout header={<div>Header</div>} footer={<div>Footer</div>}>
        <div>Content</div>
      </Layout>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('[FRAGILE] renders big contents', () => {
    const { asFragment } = render(
      <Layout header={<div>Header</div>} footer={<div>Footer</div>}>
        <div style={{ height: '110vh' }}>Content</div>
      </Layout>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders passed components', () => {
    render(
      <Layout header={<div>Header</div>} footer={<div>Footer</div>}>
        <div style={{ height: '110vh' }}>Content</div>
      </Layout>
    );

    screen.getByText(/Header/);
    screen.getByText(/Content/);
    screen.getByText(/Footer/);
  });

  it('[FRAGILE] allows to disable content padding', () => {
    const { asFragment } = render(
      <Layout header={<div>Header</div>} offPadding>
        <div>Content</div>
      </Layout>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('[FRAGILE] allows to skip footer', () => {
    const { asFragment } = render(
      <Layout header={<div>Header</div>}>
        <div>Content</div>
      </Layout>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('[FRAGILE] allows to pass sidebar and gives option to maintain it', () => {
    const { asFragment } = render(
      <Layout
        header={<div>Header</div>}
        sidebar={(toggler) => (
          <div onClick={toggler.toggle}>
            Sidebar: {toggler.opened ? 'opened' : 'closed'}
          </div>
        )}
      >
        <div>Content</div>
      </Layout>
    );

    fireEvent.click(screen.getByText(/opened/));

    expect(asFragment()).toMatchSnapshot();

    fireEvent.click(screen.getByText(/closed/));

    expect(asFragment()).toMatchSnapshot();
  });
});
