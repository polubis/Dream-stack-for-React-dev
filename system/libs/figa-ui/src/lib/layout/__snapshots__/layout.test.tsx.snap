// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User is able to use layout when [FRAGILE] allows to disable content padding 1`] = `
<DocumentFragment>
  <div
    class="sc-bcXHqe MLbvX layout off-padding"
  >
    <header
      class="layout-top-nav"
    >
      <div>
        Header
      </div>
    </header>
    <main
      class="layout-content"
    >
      <div>
        Content
      </div>
    </main>
    <footer
      class="layout-footer"
    />
    <nav
      class="layout-bottom-nav"
    >
      <div />
    </nav>
  </div>
</DocumentFragment>
`;

exports[`User is able to use layout when [FRAGILE] allows to pass sidebar and gives option to maintain it 1`] = `
<DocumentFragment>
  <div
    class="sc-bcXHqe MLbvX layout asided"
  >
    <header
      class="layout-top-nav"
    >
      <div>
        Header
      </div>
    </header>
    <main
      class="layout-content"
    >
      <aside
        class="layout-content-sidebar"
      >
        <div>
          Sidebar: closed
        </div>
      </aside>
      <div>
        Content
      </div>
    </main>
    <footer
      class="layout-footer"
    />
    <nav
      class="layout-bottom-nav"
    >
      <div />
    </nav>
  </div>
</DocumentFragment>
`;

exports[`User is able to use layout when [FRAGILE] allows to pass sidebar and gives option to maintain it 2`] = `
<DocumentFragment>
  <div
    class="sc-bcXHqe MLbvX layout asided opened"
  >
    <header
      class="layout-top-nav"
    >
      <div>
        Header
      </div>
    </header>
    <main
      class="layout-content"
    >
      <aside
        class="layout-content-sidebar"
      >
        <div>
          Sidebar: opened
        </div>
      </aside>
      <div>
        Content
      </div>
    </main>
    <footer
      class="layout-footer"
    />
    <nav
      class="layout-bottom-nav"
    >
      <div />
    </nav>
  </div>
</DocumentFragment>
`;

exports[`User is able to use layout when [FRAGILE] allows to skip footer 1`] = `
<DocumentFragment>
  <div
    class="sc-bcXHqe MLbvX layout"
  >
    <header
      class="layout-top-nav"
    >
      <div>
        Header
      </div>
    </header>
    <main
      class="layout-content"
    >
      <div>
        Content
      </div>
    </main>
    <footer
      class="layout-footer"
    />
    <nav
      class="layout-bottom-nav"
    >
      <div />
    </nav>
  </div>
</DocumentFragment>
`;

exports[`User is able to use layout when [FRAGILE] renders big contents 1`] = `
<DocumentFragment>
  <div
    class="sc-bcXHqe MLbvX layout"
  >
    <header
      class="layout-top-nav"
    >
      <div>
        Header
      </div>
    </header>
    <main
      class="layout-content"
    >
      <div
        style="height: 110vh;"
      >
        Content
      </div>
    </main>
    <footer
      class="layout-footer"
    >
      <div>
        Footer
      </div>
    </footer>
    <nav
      class="layout-bottom-nav"
    >
      <div />
    </nav>
  </div>
</DocumentFragment>
`;

exports[`User is able to use layout when [FRAGILE] renders with default setup 1`] = `
<DocumentFragment>
  <div
    class="sc-bcXHqe MLbvX layout"
  >
    <header
      class="layout-top-nav"
    >
      <div>
        Header
      </div>
    </header>
    <main
      class="layout-content"
    >
      <div>
        Content
      </div>
    </main>
    <footer
      class="layout-footer"
    >
      <div>
        Footer
      </div>
    </footer>
    <nav
      class="layout-bottom-nav"
    >
      <div>
        <div>
          Bottom nav item
        </div>
      </div>
    </nav>
  </div>
</DocumentFragment>
`;
